### ISSUE TITLE
🚨🚨 Stripe payments failing when 3D secure request kicks in

### ISSUE THREAD

**Issue by MattAppleton** [verbb/formie#1307]
### Describe the bug

We have a problem where payments to our Formie-hosted Stripe payment goes through fine for smaller amounts...

but when we try and process a larger payment it fails after the (successfully authenticated) 3D Secure 2 verification happens...

### Steps to reproduce

1. Make a card payment (in this instance for £500)

2. user gets asked for 3D Secure verification code. (at this point we see a red 'submission failed' label behind the 3D Secure panel)

3. Code gets put successfully (We then see (sometimes!) a green success message

4. But  — payments arrive on the Stripe side flagged as 'incomplete' with a message like this

"The payment pi_3McB7DJU758jprrb0pb2ozPv for £500.00 requires you to take action in order to complete the payment"
  

### Form settings

- Multi-page form: Yes or No
- Submission Method: Ajax or Page Reload
- Client-side Validation: Yes or No
- Custom Form Templates: Yes or No

### Craft CMS version

4.3.5

### Plugin version

2.021

### Multi-site?

Yes

### Additional context

We have Blitz and Barba installed.

Barba.js was giving us some trouble with form submission, but that has been ironed out (we think) with a new event handler.

```
{% set form = node.form.one() %}
<div class="w-full py-12 barbaForm" data-formid="{{ form.formId }}" data-formhandle="{{ form.handle }}">	
<script src="https://cdn.polyfill.io/v2/polyfill.js?features=fetch,Promise"></script>
```

This is causing real issues for us at the minute, if you could have a look that would be mightily appreciated!

Thank you!

**Comment by MattAppleton** [comment]
Little update, not sure if this is useful but:

Stripe tell us that the transaction has passed 3D Secure — but somehow this doesn't get passed back and the payment is registered as  incomplete' 

```3D Secure authentication succeeded
This transaction has been authenticated with 3D Secure 2. The customer was authenticated via a [challenge](https://stripe.com/guides/3d-secure-2#frictionless-authentication) flow, and shown a challenge window from their bank.
16 Feb 2023, 17:19```

**Comment by engram-design** [comment]
Fixed for the next release. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.0.22"`

**Comment by MattAppleton** [comment]
Hi @engram-design  — thanks for the quick reply... 

So you think this update will solve the 3DS issue? And our Barba setup as is is OK? Heres the form (hit apply now) on the site where the larger payments are gumming up...

https://www.tavinstitute.org/professional-development/organisational-consultancy-working-with-dynamics

All the best

**Comment by engram-design** [comment]
Yep, there was a lack of confirming the 3DS payment once approved on our end, so that should sort that out. Shouldn't have anything to do with Barba on the front-end.

**Comment by MattAppleton** [comment]
OK, getting later here but that will be first thing on the list tomorrow :)

thanks!

And — off topic I know — our client is still VERY keen to integrate Opayo payments. 

We had tried building a Formie  / Opayo method via an Upwork contact, but the whole process ended up sucking up time, £££ and not working, so we had to ditch in favour of Stripe, just for now. 

Stripe's fee structure is significantly worse for these clients than Opayo though... do let us know if you have any news on that front, and very happy to contribute towards this outcome / share what we have explored so far.


—

**Comment by engram-design** [comment]
Opayo is certainly on the [roadmap](https://github.com/verbb/formie/issues/1147), and I think the only issue other than scheduling the work, is having some testing details available. From previous looks at Sage Pay, they haven't really been able to help us with a sandbox account, being in Australia (requiring a Europe-base bank account). But that may have changed.

You're welcome to reach out directly to `web@verbb.io` to discuss it there if you like!

**Comment by MattAppleton** [comment]
Hello again @engram-design 


So we applied the update, and there is progress over yesterday's behaviour — the 3DS goes through, and the payment arrives safely with Stripe — and the form gets submitted...

but — we get an 🛑 error message — even though the payment and the form have both submitted fine...


Something awry with the timing / handling after 3DS

<img width="895" alt="Zircon-Screenshot 2023-02-17 at 10 43 47" src="https://user-images.githubusercontent.com/6295710/219648609-26d9765f-a370-47dc-bfc1-86d891536696.png">

**Comment by engram-design** [comment]
Hmmm, you're getting a request timeout. If you take a look at the network requests tab in your browsers web developer tools, how long is the XHR request actually taking? I don't _think_ the 3DS authentication stuff would alter the timeout of the XHR request.

But it sounds like the payment is going through. Was the submission created all okay?

At which specific point does this message appear? If you're able to provide a video (please use a test card!) to illustrate the flow, that'd be helpful.

**Comment by MattAppleton** [comment]
Hi @engram-design 

It all seemed pretty slow. Ill have to setup another form to test this, will have to get back to you ASAP on this, with a video :)


But to confirm

1. YES the payment went through fine — tried that twice so far
2. YES the form was correctly submitted, email notifications sent etc — despite the form hanging on the submission error at the front end...

**Comment by MattAppleton** [comment]
@engram-design  — just emailed you by the way...

**Comment by MattAppleton** [comment]
Hi @engram-design 


Sent you videos, now with the network timings.

Safari seems to do better than Chrome with this... but the Stripe + 3DS isn't in a state we can use on the form

**Comment by engram-design** [comment]
Fixed in [2.0.23](https://github.com/verbb/formie/releases/tag/2.0.23). 

I believe the outstanding issue is related to a timeout, due to having image upload fields also uploading their content at the same time. You can configure the timeout to be longer, or switch to a multi-page form, where file upload fields are on an earlier page where they are submitted first.
